### ISSUE TITLE
Only default client supports GZIP response

### ISSUE THREAD

**Issue by legendjaks** [OpenFeign/feign#1697]
The issue exists in OpenFeign 11.8, but it was working in 10.x.

Client.java
-----------
    
    
    @Override
    public Response execute(Request request, Options options) throws IOException {
      HttpURLConnection connection = convertAndSend(request, options);
      return convertResponse(connection, request);
    }

    Response convertResponse(HttpURLConnection connection, Request request) throws IOException {
      int status = connection.getResponseCode();
      String reason = connection.getResponseMessage();
      ...
      InputStream stream;
      if (status >= 400) {
        stream = connection.getErrorStream();
      } else {
        if (this.isGzip(connection.getHeaderFields().get(CONTENT_ENCODING))) {
          stream = new GZIPInputStream(connection.getInputStream());
        } else if (this.isDeflate(connection.getHeaderFields().get(CONTENT_ENCODING))) {
          stream = new InflaterInputStream(connection.getInputStream());
        } else {
          stream = connection.getInputStream();
        }
      }
     ...
    }

ApacheHttpClient.java
-----------

```

public final class ApacheHttpClient implements Client {
  private static final String ACCEPT_HEADER_NAME = "Accept";

  private final HttpClient client;
  ...

  @Override
  public Response execute(Request request, Request.Options options) throws IOException {
    HttpUriRequest httpUriRequest;
    try {
      httpUriRequest = toHttpUriRequest(request, options);
    } catch (URISyntaxException e) {
      throw new IOException("URL '" + request.url() + "' couldn't be parsed into a URI", e);
    }
    HttpResponse httpResponse = client.execute(httpUriRequest);
    return toFeignResponse(httpResponse, request);
  }

```

If any client is used other than the `default` client, the GZIP content can't be decoded and results in an exception, because GZIP is read as simple InputStream, not GZIPInputStream

```

feign.codec.DecodeException: Error while extracting response for type [java.util.List<ProductDto>] and content type [application/json]; nested exception is org.springframework.http.converter.HttpMessageNotReadableException: JSON parse error: Illegal character ((CTRL-CHAR, code 31)): only regular white space (\r, \n, \t) is allowed between tokens; nested exception is com.fasterxml.jackson.core.JsonParseException: Illegal character ((CTRL-CHAR, code 31)): only regular white space (\r, \n, \t) is allowed between tokens
```

**Comment by kdavisk6** [comment]
Feign is maintained by a series of volunteers.  We are open to Pull Requests for any issue submitted.  Please review our [Contribution Guide](https://github.com/OpenFeign/feign/blob/master/CONTRIBUTING.md) to get started.

**Comment by legendjaks** [comment]
Verified using `io.github.openfeign:feign-httpclient:11.9` client. The issue exists.

https://github.com/OpenFeign/feign/blob/83c3d69851b15384cd327b4e7fc566edd40ecb83/httpclient/src/main/java/feign/httpclient/ApacheHttpClient.java#L215

The above method has to return `GZIPInputStream` or `InflaterInputStream` if content-encoding is gzip/deflate, but it always returns `InputStream`. In the local code, changed this to `GZIPInputStream` and it worked.

**Comment by radio-rogal** [comment]
The default client has some compression tests for gzip and deflate. I've moved them to the client abstract test class.

Bunch of clients support gzip and deflate, Apache HTTP client too. As I've wrote a client just wrappes a third-party client and all "magic" to process gzip or deflate hides in this third-party code. So when the client gets gzip or deflate response it does not decompress it itself.

But some third-party libraris do not process decompression out-of-the-box.

- OkHTTP client with an interceptor can proccess GZIP response, e.g. https://stackoverflow.com/questions/51901333/okhttp-3-how-to-decompress-gzip-deflate-response-manually-using-java-android
- Google HTTP client with NetHttpTransport does not support GZIP response at all but we can use Apache HTTP client as its transport
- JAX-RS2 client does not support GZIP responses out-of-the-box but we can register a class that implies it like Jersey implementation does it, see https://stackoverflow.com/questions/22519642/how-to-enable-gzip-compression-for-content-encoding-with-jersey-jax-rs-2-0-cli
